plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -204,7 +204,13 @@ InputStream readBlob(String blobName) throws IOException {
                     Storage.Objects.Get object = client.objects().get(bucket, blobName);
                     return object.executeMediaAsInputStream();
                 });
-            }catch(Exception e){
+            }catch (GoogleJsonResponseException e) {
+                if(e.getStatusCode() == 404){
+                    throw new NoSuchFileException(e.getMessage());
+                } else {
+                    throw e;
```

Comment:
Do we not want this to feed into the retry logic? Is it other exception types that we get (below) where retries are appropriate?

Reply:
The only problem was that Google changed the way it sends back the exception for 404 not found. Elasticsearch was expecting a NoSuchFileException in order to initialize the new snapshot/repository, but in the current 5.6-dr the propagated exception was a different type so it would fail to create the required files. This addition to the code is in the retry mechanism . We'll do more tests on it of course, but this was the scenario that was failing

Reply:
I'm just wondering if there are other non-404 GoogleJsonResponseExceptions being thrown, that previously might have been retried and worked. Maybe not?

Reply:
ah, no, I'm actually checking for the return code , 404. this will always be File not found, no matter how it looks, so we should be fully covered here.

Reply:
OK, so if it is not a 404, there is no hope of retrying, so the appropriate thing to do is to just throw the exception. Right?

Reply:
It still retries, but it throws the received exception from the underlying layer. If you check the code, the try catch is nested into the retry mechanism

Reply:
As mentioned, there is still some due diligence to be done here, but except for the fact that the initial issue was resolved the code runs exactly under the same logic

Reply:
It won't retry if there is a GoogleJsonResponseException with a response code of (for example) 503. It'll throw the GoogleJsonResponseException and bail out of this method. That's a change in behavior from before. I have no idea if that actually ever happens, or if so if it is ever worth retrying though.

Reply:
Actually that's correct, we haven't encountered any other situation but to keep it accurate, I'll update it so it reflects the original version

Reply:
modified it